File: pkg/security/permissions.go
// Package security provides fast permission checking for agent actions.
package security

import (
	"os/user"
	"path/filepath"
	"regexp"
	"strings"
)

// RiskLevel represents the severity of an action.
type RiskLevel int

//go:generate go run golang.org/x/tools/cmd/stringer -type=RiskLevel -linecomment

const (
	RiskSafe RiskLevel = iota   // SAFE
	RiskLow                       // LOW
	RiskMedium                    // MEDIUM
	RiskHigh                      // HIGH
	RiskCritical                  // CRITICAL
)

// ActionRule describes permissions for an action type.
type ActionRule struct {
	Action               string
	RiskLevel            RiskLevel
	RequiresConfirmation bool
}

// Builtin action rules.
var BuiltinRules = map[string]ActionRule{
	// File operations
	"file_read":   {Action: "file_read", RiskLevel: RiskSafe, RequiresConfirmation: false},
	"file_write":  {Action: "file_write", RiskLevel: RiskMedium, RequiresConfirmation: false},
	"file_delete": {Action: "file_delete", RiskLevel: RiskHigh, RequiresConfirmation: true},

	// Shell/system operations
	"shell_execute":   {Action: "shell_execute", RiskLevel: RiskMedium, RequiresConfirmation: false},
	"install_package": {Action: "install_package", RiskLevel: RiskHigh, RequiresConfirmation: true},
	"system_modify":   {Action: "system_modify", RiskLevel: RiskCritical, RequiresConfirmation: true},

	// Network operations
	"network_request": {Action: "network_request", RiskLevel: RiskLow, RequiresConfirmation: false},
	"send_message":    {Action: "send_message", RiskLevel: RiskMedium, RequiresConfirmation: false},

	// Memory operations
	"memory_read":  {Action: "memory_read", RiskLevel: RiskSafe, RequiresConfirmation: false},
	"memory_write": {Action: "memory_write", RiskLevel: RiskLow, RequiresConfirmation: false},

	// Platform introspection (always safe - read-only)
	"platform_read": {Action: "platform_read", RiskLevel: RiskSafe, RequiresConfirmation: false},

	// Task management
	"task_read":  {Action: "task_read", RiskLevel: RiskSafe, RequiresConfirmation: false},
	"task_write": {Action: "task_write", RiskLevel: RiskLow, RequiresConfirmation: false},

	// Agent operations
	"agent_delegate": {Action: "agent_delegate", RiskLevel: RiskLow, RequiresConfirmation: false},
	"skill_execute":  {Action: "skill_execute", RiskLevel: RiskMedium, RequiresConfirmation: false},
}

// Dangerous command patterns that elevate shell_execute to HIGH risk.
var dangerousCommandsRE = regexp.MustCompile(
	`(?i)\b(rm\s+-rf|mkfs|dd\s+if=|chmod\s+-R|chown\s+-R|shutdown|reboot` +
		`|init\s+[06]|systemctl\s+(stop|disable|mask)|kill\s+-9` +
		`|iptables|nft|deluser|userdel|groupdel)\b`,
)

// Financial operation patterns.
var financialPatternsRE = regexp.MustCompile(
	`(?i)\b(transfer\s+(funds?|money|payment)|send\s+(payment|money|funds?)` +
		`|wire\s+transfer|purchase|buy|sell|trade|withdraw` +
		`|credit\s*card|bank\s*account|routing\s*number` +
		`|cryptocurrency|bitcoin|ethereum|wallet\s*address)\b`,
)

// Config holds permission checker configuration.
type Config struct {
	AllowedPaths                []string
	BlockedPaths                []string
	BlockFinancial              bool
	RequireConfirmationHigh     bool
	RequireConfirmationCritical bool
}

// PermissionChecker provides fast permission checking.
type PermissionChecker struct {
	config       Config
	allowedGlobs []string
	blockedGlobs []string
	homeDir      string
}

// NewPermissionChecker creates a new permission checker.
func NewPermissionChecker(cfg Config) *PermissionChecker {
	pc := &PermissionChecker{
		config:       cfg,
		allowedGlobs: make([]string, 0, len(cfg.AllowedPaths)),
		blockedGlobs: make([]string, 0, len(cfg.BlockedPaths)),
	}

	// Get home directory for tilde expansion
	if u, err := user.Current(); err == nil {
		pc.homeDir = u.HomeDir
	}

	// Pre-expand paths
	for _, p := range cfg.AllowedPaths {
		pc.allowedGlobs = append(pc.allowedGlobs, pc.expandPath(p))
	}
	for _, p := range cfg.BlockedPaths {
		pc.blockedGlobs = append(pc.blockedGlobs, pc.expandPath(p))
	}

	return pc
}

func (pc *PermissionChecker) expandPath(path string) string {
	if strings.HasPrefix(path, "~") && pc.homeDir != "" {
		path = pc.homeDir + path[1:]
	}
	// Clean the path
	return filepath.Clean(path)
}

// CheckPath returns true if the path is allowed.
func (pc *PermissionChecker) CheckPath(path string) bool {
	allowed, _ := pc.checkPathWithReason(path)
	return allowed
}

// CheckPathWithReason returns (allowed, reason).
func (pc *PermissionChecker) checkPathWithReason(pathStr string) (ok bool, result string) {
	resolved := pc.expandPath(pathStr)
	if absPath, err := filepath.Abs(resolved); err == nil {
		resolved = absPath
	}

	// Block list takes precedence
	for _, pattern := range pc.blockedGlobs {
		if pc.matchGlobPattern(pattern, resolved) {
			return false, "Path matches blocked pattern: " + pattern
		}
	}

	// If there's an allow list, path must match at least one
	if len(pc.allowedGlobs) > 0 {
		for _, pattern := range pc.allowedGlobs {
			if pc.matchGlobPattern(pattern, resolved) {
				return true, "Path is within allowed paths"
			}
		}
		return false, "Path does not match any allowed path pattern"
	}

	return true, "No path restrictions configured"
}

// matchGlobPattern handles both exact globs and directory prefixes.
// Patterns ending with /* or /** are treated as directory prefixes.
func (pc *PermissionChecker) matchGlobPattern(pattern, path string) bool {
	// Handle trailing /* or /** as directory prefix match
	if strings.HasSuffix(pattern, "/*") || strings.HasSuffix(pattern, "/**") {
		// Strip the trailing /* or /**
		dirPrefix := strings.TrimSuffix(strings.TrimSuffix(pattern, "**"), "*")
		dirPrefix = strings.TrimSuffix(dirPrefix, "/")

		// Check if path equals the directory or is under it
		if path == dirPrefix {
			return true
		}
		// Check if path is under this directory
		if strings.HasPrefix(path, dirPrefix+"/") {
			return true
		}
		return false
	}

	// Try exact filepath.Match for non-prefix patterns
	if matched, _ := filepath.Match(pattern, path); matched {
		return true
	}

	// Also check exact path match (for patterns without wildcards)
	if pattern == path {
		return true
	}

	// Check if path is under the pattern directory (for directory patterns without *)
	if strings.HasPrefix(path, pattern+"/") {
		return true
	}

	return false
}

// EvaluateShellRisk returns the risk level for a shell command.
func EvaluateShellRisk(command string) RiskLevel {
	if dangerousCommandsRE.MatchString(command) {
		return RiskHigh
	}
	return RiskMedium
}

// IsFinancial returns true if the text contains financial operation patterns.
func IsFinancial(text string) bool {
	return financialPatternsRE.MatchString(text)
}

// CheckResult is the result of a permission check.
type CheckResult struct {
	Allowed       bool
	Reason        string
	EffectiveRisk RiskLevel
	NeedsConfirm  bool
}

// CheckPermission checks if an action is permitted.
func (pc *PermissionChecker) CheckPermission(action string, details map[string]string) CheckResult {
	// Look up base rule
	rule, ok := BuiltinRules[action]
	if !ok {
		return CheckResult{
			Allowed: false,
			Reason:  "Unknown action: " + action,
		}
	}

	effectiveRisk := rule.RiskLevel

	// Financial gate
	if pc.config.BlockFinancial {
		for _, v := range details {
			if IsFinancial(v) {
				return CheckResult{
					Allowed: false,
					Reason:  "Financial operations are blocked by policy",
				}
			}
		}
	}

	// Path-based checks for file actions
	if action == "file_read" || action == "file_write" || action == "file_delete" {
		if path, ok := details["path"]; ok && path != "" {
			allowed, reason := pc.checkPathWithReason(path)
			if !allowed {
				return CheckResult{
					Allowed: false,
					Reason:  reason,
				}
			}
		}
	}

	// Shell command risk elevation
	if action == "shell_execute" {
		if command, ok := details["command"]; ok {
			effectiveRisk = EvaluateShellRisk(command)
		}
	}

	// Confirmation gating
	needsConfirm := effectiveRisk >= RiskHigh && pc.config.RequireConfirmationHigh

	if effectiveRisk >= RiskCritical && pc.config.RequireConfirmationCritical {
		needsConfirm = true
	}

	if needsConfirm {
		return CheckResult{
			Allowed:       false,
			Reason:        "Action requires user confirmation",
			EffectiveRisk: effectiveRisk,
			NeedsConfirm:  true,
		}
	}

	return CheckResult{
		Allowed:       true,
		Reason:        "Permitted",
		EffectiveRisk: effectiveRisk,
	}
}

File: archive/python-legacy/meept/security/permissions.py
"""Action permission gating for the autonomous agent.

Every action the agent attempts (file I/O, shell commands, network
requests, ...) is routed through the :class:`PermissionManager` which
decides whether to allow, deny, or request human confirmation based on:

* a risk-level classification for each action type,
* path allow/block lists from the security configuration, and
* special-case rules for financial operations.

HIGH and CRITICAL actions can optionally require explicit user
confirmation via an :class:`asyncio.Future` handshake.
"""

from __future__ import annotations

import asyncio
import enum
import fnmatch
import logging
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from meept.models.config_schema import SecurityConfig

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Risk classification
# ---------------------------------------------------------------------------


class RiskLevel(enum.IntEnum):
    """Severity tiers for agent-initiated actions.

    Uses :class:`~enum.IntEnum` so that comparisons (``>=``, ``<``, etc.)
    work naturally without custom dunder methods.
    """

    SAFE = 0
    LOW = 1
    MEDIUM = 2
    HIGH = 3
    CRITICAL = 4


@dataclass(slots=True)
class ActionPermission:
    """Descriptor for a single permitted action category."""

    action: str
    risk_level: RiskLevel
    description: str
    requires_confirmation: bool


# ---------------------------------------------------------------------------
# Built-in action rules
# ---------------------------------------------------------------------------

# Base risk table.  The PermissionManager may dynamically elevate risk for
# specific invocations (e.g. ``shell_execute`` with ``rm -rf``).

_BUILTIN_RULES: dict[str, ActionPermission] = {
    "file_read": ActionPermission(
        action="file_read",
        risk_level=RiskLevel.SAFE,
        description="Read a file from the filesystem",
        requires_confirmation=False,
    ),
    "file_write": ActionPermission(
        action="file_write",
        risk_level=RiskLevel.MEDIUM,
        description="Write or overwrite a file on the filesystem",
        requires_confirmation=False,
    ),
    "file_delete": ActionPermission(
        action="file_delete",
        risk_level=RiskLevel.HIGH,
        description="Permanently delete a file from the filesystem",
        requires_confirmation=True,
    ),
    "shell_execute": ActionPermission(
        action="shell_execute",
        risk_level=RiskLevel.MEDIUM,
        description="Execute a shell command",
        requires_confirmation=False,
    ),
    "network_request": ActionPermission(
        action="network_request",
        risk_level=RiskLevel.LOW,
        description="Make an outbound HTTP/HTTPS request",
        requires_confirmation=False,
    ),
    "send_message": ActionPermission(
        action="send_message",
        risk_level=RiskLevel.MEDIUM,
        description="Send a message to a user or external service",
        requires_confirmation=False,
    ),
    "install_package": ActionPermission(
        action="install_package",
        risk_level=RiskLevel.HIGH,
        description="Install a software package on the system",
        requires_confirmation=True,
    ),
    "system_modify": ActionPermission(
        action="system_modify",
        risk_level=RiskLevel.CRITICAL,
        description="Modify system-level configuration or settings",
        requires_confirmation=True,
    ),
}

# Shell commands that bump ``shell_execute`` to HIGH.
_DANGEROUS_COMMANDS_RE = re.compile(
    r"\b(rm\s+-rf|mkfs|dd\s+if=|chmod\s+-R|chown\s+-R|shutdown|reboot"
    r"|init\s+[06]|systemctl\s+(stop|disable|mask)|kill\s+-9"
    r"|iptables|nft|deluser|userdel|groupdel)\b",
    re.IGNORECASE,
)

# Patterns indicating a financial operation (always blocked when
# ``block_financial`` is set).
_FINANCIAL_PATTERNS_RE = re.compile(
    r"\b(transfer\s+(funds?|money|payment)|send\s+(payment|money|funds?)"
    r"|wire\s+transfer|purchase|buy|sell|trade|withdraw"
    r"|credit\s*card|bank\s*account|routing\s*number"
    r"|cryptocurrency|bitcoin|ethereum|wallet\s*address)\b",
    re.IGNORECASE,
)


# ---------------------------------------------------------------------------
# PermissionManager
# ---------------------------------------------------------------------------


class PermissionManager:
    """Central authority for agent action gating.

    Parameters
    ----------
    config:
        The ``[security]`` section of the application config.
    """

    def __init__(self, config: SecurityConfig) -> None:
        self._config = config

        # Pre-expand glob patterns for path matching.
        self._allowed_globs: list[str] = [
            str(Path(p).expanduser()) for p in config.allowed_paths
        ]
        self._blocked_globs: list[str] = [
            str(Path(p).expanduser()) for p in config.blocked_paths
        ]

        # Pending confirmation futures keyed by a unique request id.
        self._pending: dict[str, asyncio.Future[bool]] = {}
        self._next_id = 0

    # -- Path checking ------------------------------------------------------

    def _is_path_allowed(self, path_str: str) -> tuple[bool, str]:
        """Check *path_str* against allow/block lists.

        Returns ``(allowed, reason)``.
        """
        resolved = str(Path(path_str).expanduser().resolve())

        # Block list takes precedence.
        for pattern in self._blocked_globs:
            if fnmatch.fnmatch(resolved, pattern):
                return False, f"Path matches blocked pattern: {pattern}"

        # If there is an allow list, the path must match at least one entry.
        if self._allowed_globs:
            for pattern in self._allowed_globs:
                if fnmatch.fnmatch(resolved, pattern):
                    return True, "Path is within allowed paths"
            return False, "Path does not match any allowed path pattern"

        return True, "No path restrictions configured"

    def check_path(self, path: str) -> bool:
        """Return ``True`` if *path* is within allowed paths and not blocked.

        Parameters
        ----------
        path:
            Absolute or ``~``-prefixed filesystem path to check.
        """
        allowed, _reason = self._is_path_allowed(path)
        return allowed

    # -- Shell command risk elevation ---------------------------------------

    @staticmethod
    def _evaluate_shell_risk(command: str) -> RiskLevel:
        """Return the effective risk level for a shell command."""
        if _DANGEROUS_COMMANDS_RE.search(command):
            return RiskLevel.HIGH
        return RiskLevel.MEDIUM

    # -- Financial detection ------------------------------------------------

    @staticmethod
    def _is_financial(details: dict[str, Any]) -> bool:
        """Return ``True`` if *details* describe a financial operation."""
        for value in details.values():
            if isinstance(value, str) and _FINANCIAL_PATTERNS_RE.search(value):
                return True
        return False

    # -- Public API ---------------------------------------------------------

    def check_permission(
        self, action: str, details: dict[str, Any] | None = None,
    ) -> tuple[bool, str]:
        """Decide whether *action* with *details* is permitted.

        Returns
        -------
        tuple[bool, str]
            ``(allowed, reason)``
        """
        details = details or {}

        # Look up the base rule.
        rule = _BUILTIN_RULES.get(action)
        if rule is None:
            logger.warning("Unknown action requested: %s", action)
            return False, f"Unknown action: {action}"

        effective_risk = rule.risk_level

        # --- Financial gate ---
        if self._config.block_financial and self._is_financial(details):
            logger.warning("Blocked financial action: %s %s", action, details)
            return False, "Financial operations are blocked by policy"

        # --- Path-based checks for file actions ---
        if action in ("file_read", "file_write", "file_delete"):
            path = details.get("path", "")
            if path:
                allowed, reason = self._is_path_allowed(path)
                if not allowed:
                    logger.warning(
                        "Path denied for %s: %s (%s)", action, path, reason,
                    )
                    return False, reason

        # --- Shell command risk elevation ---
        if action == "shell_execute":
            command = details.get("command", "")
            effective_risk = self._evaluate_shell_risk(command)

        # --- Confirmation gating ---
        needs_confirm = (
            (effective_risk >= RiskLevel.HIGH and self._config.require_confirmation_high)
            or (
                effective_risk >= RiskLevel.CRITICAL
                and self._config.require_confirmation_critical
            )
        )

        if needs_confirm:
            logger.info(
                "Action %s (risk=%s) requires user confirmation",
                action,
                effective_risk.name,
            )
            return False, (
                f"Action '{action}' has risk level {effective_risk.name} and requires "
                f"user confirmation before execution"
            )

        logger.debug("Action %s permitted (risk=%s)", action, effective_risk.name)
        return True, "Permitted"

    # -- Async confirmation handshake ---------------------------------------

    def request_confirmation(
        self, action: str, details: dict[str, Any] | None = None,
    ) -> tuple[str, asyncio.Future[bool]]:
        """Create a confirmation request for a high-risk action.

        Returns
        -------
        tuple[str, asyncio.Future[bool]]
            A unique request ID and a future that resolves to ``True``
            (confirmed) or ``False`` (denied) once the human responds.
        """
        loop = asyncio.get_running_loop()
        future: asyncio.Future[bool] = loop.create_future()

        request_id = f"confirm-{self._next_id}"
        self._next_id += 1
        self._pending[request_id] = future

        logger.info(
            "Confirmation requested for %s (id=%s, details=%s)",
            action,
            request_id,
            details,
        )
        return request_id, future

    def resolve_confirmation(self, request_id: str, *, approved: bool) -> None:
        """Resolve a pending confirmation.

        Parameters
        ----------
        request_id:
            The ID returned by :meth:`request_confirmation`.
        approved:
            ``True`` to allow the action, ``False`` to deny.
        """
        future = self._pending.pop(request_id, None)
        if future is None:
            logger.warning("No pending confirmation for id=%s", request_id)
            return
        if not future.done():
            future.set_result(approved)
            logger.info(
                "Confirmation %s resolved: %s",
                request_id,
                "approved" if approved else "denied",
            )
